apps/web/src/services/dashboards/shared/dashboard-widget-input-form/dashboard-widget-options-form/DashboardWidgetOptionDropdown.vue

```diff
@@ -2,129 +2,311 @@
+        variableModels.push(variableModel);
+    });
+    return variableModels;
 };
-const handleUpdateSelected = (selected: SelectDropdownMenuItem[]) => {
-    emit('update:selected', selected);
+const getInheritOptionMenuHandler = (schema?: WidgetOptionsSchemaProperty): AutocompleteHandler => {
+    if (!schema) return () => ({ results: [] });
+
+    const selectableVariableMenuItems: {name: string; label: string;}[] = [];
+    Object.entries(props.variablesSchema?.properties ?? {}).forEach(([propertyName, property]) => {
+        if (property.use && property.selection_type === schema.selection_type) {
+            selectableVariableMenuItems.push({ name: getWidgetOptionKeyByVariableKey(propertyName), label: property.name });
+        }
+    });
+    return (inputText) => {
+        const regex = getTextHighlightRegex(inputText);
+        const filteredMenuItems = selectableVariableMenuItems.filter((item) => regex.test(item.label));
+        return {
+            results: filteredMenuItems,
+        };
+    };
+};
+// refine selected item by handlers (e.g. [{ name: 'aws' }] -> [{ name: 'aws', label: 'AWS' }])
+const getRefinedSelectedItemByHandlers = async (
+    handlers: AutocompleteHandler[],
+    selected: SelectDropdownMenuItem|SelectDropdownMenuItem[],
+): Promise<SelectDropdownMenuItem[]> => {
+    if (!selected) return [];
+    let _selected = Array.isArray(selected) ? selected : [selected];
+    const promiseResults = await Promise.allSettled(handlers.map((handler) => handler(
+        '',
+        undefined,
+        undefined,
+        _selected,
+    )));
+    promiseResults.forEach((result, idx) => {
+        if (result.status === 'fulfilled') {
+            const results = result.value.results;
+            _selected = _selected.map((item) => {
+                const found = results.find((d) => d.name === item.name);
+                if (found) return found;
+                return item;
+            });
+        } else {
+            console.error(new Error(`Failed to fetch data from handler: ${idx}`));
+        }
+    });
+    return _selected;
 };
+// init selected menu items on init or inherit changed
+const initSelectedMenuItems = async (): Promise<SelectDropdownMenuItem[]> => {
+    const inheritOption = widgetFormStore.inheritOptions?.[props.propertyName];
 
+    // 1) inherit case
+    if (state.inherit) {
+        if (!state.inheritanceMode || state.inheritanceMode === 'KEY_MATCHING') {
+            const variableKey = inheritOption.variable_key;
+            const variableSchema = props.variablesSchema?.properties?.[variableKey];
+            if (!variableKey || !variableSchema?.use) return [];
+            return [{ name: variableKey, label: variableSchema.name }];
+        }
+        return [];
+    }
 
-watch(() => props.inherit, () => {
-    state.showErrorMessage = false;
-});
+    // 2) non-inherit case
+    const selected = get(widgetFormState.widgetOptions, props.propertyName);
```

Comment:
suggestion: It is far better if there is type declaration that `selected` can be.
For example: 
```suggestion
    const selected: WidgetFilter[]|string|undefined = get(widgetFormState.widgetOptions, props.propertyName);
```